File: components/radio/android/CiataRadioGroup.kt
package org.ciata.ds.radio

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

data class CiataRadioOption(
    val value: String,
    val label: String,
    val enabled: Boolean = true,
)

/** Implementação experimental do CMP-0004 Radio para Jetpack Compose. */
@Composable
fun CiataRadioGroup(
    legend: String,
    options: List<CiataRadioOption>,
    selectedValue: String?,
    onSelected: (String) -> Unit,
    modifier: Modifier = Modifier,
    required: Boolean = false,
    enabled: Boolean = true,
    helpText: String? = null,
    errorText: String? = null,
) {
    val normalizedLegend = legend.trim()
    val normalizedOptions = options.map { option ->
        option.copy(label = option.label.trim())
    }
    val normalizedHelpText = helpText?.trim()
    val normalizedErrorText = errorText?.trim()

    require(normalizedLegend.isNotEmpty()) { "legend não pode ser vazio." }
    require(normalizedOptions.size >= 2) { "Radio Group deve possuir ao menos duas opções." }
    require(normalizedOptions.all { it.label.isNotEmpty() }) { "Rótulos não podem ser vazios." }
    require(normalizedOptions.map { it.value }.distinct().size == normalizedOptions.size) { "Valores devem ser únicos." }
    require(selectedValue == null || normalizedOptions.any { it.value == selectedValue }) {
        "selectedValue não pertence às opções do grupo."
    }

    val groupModifier = if (!normalizedErrorText.isNullOrEmpty()) {
        modifier
            .selectableGroup()
            .semantics { error(normalizedErrorText) }
    } else {
        modifier.selectableGroup()
    }

    Column(modifier = groupModifier) {
        Text(if (required) "$normalizedLegend (obrigatório)" else normalizedLegend)

        normalizedOptions.forEach { option ->
            val optionEnabled = enabled && option.enabled
            Row(
                modifier = Modifier.selectable(
                    selected = selectedValue == option.value,
                    enabled = optionEnabled,
                    role = Role.RadioButton,
                    onClick = { onSelected(option.value) },
                ),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                RadioButton(
                    selected = selectedValue == option.value,
                    onClick = null,
                    enabled = optionEnabled,
                )
                Spacer(Modifier.width(8.dp))
                Text(option.label)
            }
        }

        if (!normalizedHelpText.isNullOrEmpty()) {
            Text(normalizedHelpText)
        }
        if (!normalizedErrorText.isNullOrEmpty()) {
            Text(normalizedErrorText)
        }
    }
}

File: components/radio/ios/CiataRadioGroup.swift
import SwiftUI

public struct CiataRadioOption: Identifiable, Hashable {
    public let id: String
    public let label: String
    public let enabled: Bool

    public init(value: String, label: String, enabled: Bool = true) {
        let normalizedLabel = label.trimmingCharacters(in: .whitespacesAndNewlines)
        precondition(!normalizedLabel.isEmpty, "label não pode ser vazio")
        self.id = value
        self.label = normalizedLabel
        self.enabled = enabled
    }
}

/// Implementação experimental do CMP-0004 Radio para SwiftUI.
public struct CiataRadioGroup: View {
    private let legend: String
    private let options: [CiataRadioOption]
    @Binding private var selectedValue: String?
    private let required: Bool
    private let disabled: Bool
    private let helpText: String?
    private let errorText: String?

    public init(
        legend: String,
        options: [CiataRadioOption],
        selectedValue: Binding<String?>,
        required: Bool = false,
        disabled: Bool = false,
        helpText: String? = nil,
        errorText: String? = nil
    ) {
        let normalizedLegend = legend.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedHelpText = helpText?.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalizedErrorText = errorText?.trimmingCharacters(in: .whitespacesAndNewlines)

        precondition(!normalizedLegend.isEmpty, "legend não pode ser vazio")
        precondition(options.count >= 2, "Radio Group deve possuir ao menos duas opções")
        precondition(Set(options.map(\.id)).count == options.count, "Valores devem ser únicos")
        precondition(selectedValue.wrappedValue == nil || options.contains { $0.id == selectedValue.wrappedValue }, "selectedValue não pertence às opções do grupo")

        self.legend = normalizedLegend
        self.options = options
        self._selectedValue = selectedValue
        self.required = required
        self.disabled = disabled
        self.helpText = normalizedHelpText
        self.errorText = normalizedErrorText
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(required ? "\(legend) (obrigatório)" : legend)
                .fontWeight(.semibold)

            Picker(legend, selection: $selectedValue) {
                ForEach(options) { option in
                    Text(option.label)
                        .tag(Optional(option.id))
                        .disabled(!option.enabled)
                }
            }
            .pickerStyle(.inline)
            .disabled(disabled)
            .accessibilityHint(accessibilityHint)

            if let helpText, !helpText.isEmpty {
                Text(helpText)
                    .font(.footnote)
                    .foregroundStyle(.secondary)
            }

            if let errorText, !errorText.isEmpty {
                Text(errorText)
                    .font(.footnote)
                    .accessibilityLabel("Erro: \(errorText)")
            }
        }
    }

    private var accessibilityHint: String {
        [helpText, errorText.map { "Erro: \($0)" }]
            .compactMap { $0 }
            .filter { !$0.isEmpty }
            .joined(separator: ". ")
    }
}
